### PR TITLE
Add "DISTRIBUTED_ONLY" deployment mode support to NSX-T Edge Gateway

### LINKED ISSUE #1299: vcd_nsxt_edgegateway with distributed only
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Allow to create routed gateway with Distributed only

"Edge Gateways (backed by Tier-1 NSX GWs) can now also be deployed in the Distributed only mode without the SR (services router) component. In such case the Tier-1 GW does not provide services that run on SR such as firewalling, NAT, VPN, DNS forwarding or static routes. The distributed nature guarantees high N/S data throughput (no hairpinning of traffic to single edge node running the active SR component). Load balancing, rate limiting or DHCP service in network mode is still supported. The other use case beside performance is to limit the NSX licensing costs."

alt text

### New or Affected Resource(s)

* vcd_nsxt_edgegateway
* vcd_nsxt_firewall (maybe restrict?)
* vcd_nsxt_ipsec_vpn_tunnel (maybe restrict?)
* vcd_nsxt_nat_rule (maybe restrict?)
* vcd_nsxt_edgegateway_static_route (maybe restrict?)
* vcd_nsxt_edgegateway_dns (maybe restrict?)

### References

na

### PR DESCRIPTION
**Pull request by Didainius**
Closes #1299 

This PR adds new field - `deployment_mode` that supports two values `ACTIVE_STANDBY` (the default that the edge gateway worked always) and `DISTRIBUTED_ONLY` . A new feature of VCD 10.6

### NON-DISTRIBUTED Edge Gateway
Edge Gateways (backed by Tier-1 NSX GWs) can now also be deployed in the Distributed only mode without the SR (services router) component. In such case the Tier-1 GW does not provide services that run on SR such as firewalling, NAT, VPN, DNS forwarding or static routes. The distributed nature guarantees high N/S data throughput (no hairpinning of traffic to single edge node running the active SR component). Load balancing, rate limiting or DHCP service in network mode is still supported. The other use case beside performance is to limit the NSX licensing costs.

Edge Gateway in `NON_DISTRIBUTED` mode supports only a small subset of configurations. The error is not completelly clear therefore there is an additional effort to make the error more clear if it matches the error for known resources that fail.

```
Error: [nsx-t firewall create/update] error creating NSX-T Firewall Rules: error setting
NSX-T Firewall: error in HTTP PUT request: ACCESS_TO_RESOURCE_IS_FORBIDDEN - [
19-2024-08-05-15-02-04-250--9e6beec5-5b47-4797-ab0d-162fed8d1401 ] Either you need some or
all of the following rights [ORG_VDC_GATEWAY_VIEW_FIREWALL] to perform operations
[GATEWAY_VIEW_FIREWALL_NSX_T] for 5f1fc518-865a-4c43-8b13-408c11ed8c06 or the target entity
is invalid.
```

to 

```
│ Error: error setting NSX-T Firewall: error in HTTP PUT request: ACCESS_TO_RESOURCE_IS_FORBIDDEN - [ 19-2024-08-06-11-14-20-775--09cd9edc-5d2f-458d-9a98-608a3d178004 ] Either you need some or all of the following rights [ORG_VDC_GATEWAY_VIEW_FIREWALL] to perform operations [GATEWAY_VIEW_FIREWALL_NSX_T] for ccf22499-c009-47ab-9472-a05dfba80391 or the target entity is invalid.
│ 
│ vcd_nsxt_firewall cannot be configured on DISTRIBUTED_ONLY NSX-T Edge Gateway
│ 
│   with vcd_nsxt_firewall.testing,
│   on vcd.TestAccVcdNsxtEdgeGateway.tf line 52, in resource "vcd_nsxt_firewall" "testing":
│   52: resource "vcd_nsxt_firewall" "testing" {
```

Tested on 10.4.0 and 10.6.0 (nsxt tag)

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -69,3 +69,5 @@ require (
 	google.golang.org/grpc v1.63.2 // indirect
```
Comment (adambarreiro) [inline go.mod:69]: Missing changelog

File: vcd/remove_leftovers_test.go [hunk 2]
```diff
@@ -525,7 +526,7 @@ func removeLeftovers(govcdClient *govcd.VCDClient, verbose bool) error {
 	// --------------------------------------------------------------
 	// External Endpoints
 	// --------------------------------------------------------------
-	if govcdClient.Client.IsSysAdmin {
+	if govcdClient.Client.IsSysAdmin && govcdClient.Client.APIVCDMaxVersionIs(">= 37.3") {
```
Comment (adambarreiro) [inline vcd/remove_leftovers_test.go:529]: Thanks for fixing 😄

**Review by adambarreiro — COMMENTED** [review]
Quick scan

File: vcd/resource_vcd_nsxt_edgegateway.go [hunk 3]
```diff
@@ -156,6 +156,14 @@ func resourceVcdNsxtEdgeGateway() *schema.Resource {
 				Optional:    true,
 				Description: "Dedicating the External Network will enable Route Advertisement for this Edge Gateway.",
 			},
+			"deployment_mode": {
```
Comment (adambarreiro) [inline vcd/resource_vcd_nsxt_edgegateway.go:159]: I wonder if this will make `vcd.ResourceSchema-vcd_nsxt_edgegateway.tf` upgrade test fail?

Comment (Didainius) [reply vcd/resource_vcd_nsxt_edgegateway.go:159]: Yes! Will add

**Review by adambarreiro — APPROVED** [review]

**Review by lvirbalas — CHANGES_REQUESTED** [review]
I can't seem to find docs for this feature in the PR :)

File: vcd/resource_vcd_nsxt_edgegateway.go (hunk 3 above)
Comment (Didainius) [reply vcd/resource_vcd_nsxt_edgegateway.go:159]: Added

**Comment by Didainius** [comment]
> I can't seem to find docs for this feature in the PR :)

Indeed. Added

File: website/docs/r/nsxt_edgegateway.html.markdown [hunk 4]
```diff
@@ -333,6 +333,10 @@ definition at provider level.
 can be used to lookup ID by name.
 * `edge_cluster_id` - (Optional) Specific Edge Cluster ID if required
 * `dedicate_external_network` - (Optional) Dedicating the external network will enable Route Advertisement for this Edge Gateway. Default `false`.
+* `deployment_mode` - (Optional, *v3.14+*,*VCD 10.6+*) `ACTIVE_STANDBY` (default) or
```
Comment (lvirbalas) [inline website/docs/r/nsxt_edgegateway.html.markdown:336]: Nit:
```suggestion
* `deployment_mode` - (Optional, *v3.14+*, *VCD 10.6+*) `ACTIVE_STANDBY` (default) or
```

**Review by lvirbalas — APPROVED** [review]
Just a tiny polish comment.

File: website/docs/r/nsxt_edgegateway.html.markdown (hunk 4 above)
Comment (Didainius) [reply website/docs/r/nsxt_edgegateway.html.markdown]: Changed here and in other place :)

**Pull request merged**